scylla-cql/src/frame/request/query.rs

```diff
@@ -204,3 +210,58 @@ impl<'q> QueryParameters<'q> {
         })
     }
 }
+
+#[derive(Debug, Clone)]
+pub enum PagingState {
+    HasMorePages {
+        next_page_handle: PagingContinuation,
+    },
+    NoMorePages,
+}
+
+impl PagingState {
+    #[inline]
+    pub fn finished(&self) -> bool {
+        matches!(*self, Self::NoMorePages)
+    }
+
+    pub(crate) fn new_from_raw_bytes(raw_paging_state: Option<&[u8]>) -> Self {
+        match raw_paging_state {
+            Some(raw_bytes) => Self::HasMorePages {
+                next_page_handle: PagingContinuation::new_from_raw_bytes(raw_bytes),
+            },
+            None => Self::NoMorePages,
+        }
+    }
+
+    #[inline]
+    pub fn into_paging_continuation_opt(self) -> Option<PagingContinuation> {
+        match self {
+            Self::HasMorePages { next_page_handle } => Some(next_page_handle),
+            Self::NoMorePages => None,
+        }
+    }
+
+    #[inline]
+    pub fn into_paging_control_flow(self) -> ControlFlow<(), PagingContinuation> {
+        match self {
+            Self::HasMorePages { next_page_handle } => ControlFlow::Continue(next_page_handle),
+            Self::NoMorePages => ControlFlow::Break(()),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PagingContinuation(Arc<[u8]>);
+
+impl PagingContinuation {
+    #[inline]
+    pub(crate) fn new_from_raw_bytes(raw_bytes: &[u8]) -> Self {
+        Self(Arc::from(raw_bytes))
+    }
+
+    #[inline]
+    pub fn as_byte_slice(&self) -> &[u8] {
+        self.0.deref()
+    }
+}
```

Comment:
I have to admit that I'm not exactly the fan of this API:
- The naming isn't great: `PagingContinuation` / `into_paging_continuation_opt` are imo quite unfriendly names with not a very obvious meaning.
- `into_paging_continuation_opt` returns a value that also represents a beggining of paging query. Consider the example from session_test:
```rust
        let mut paging_continuation: Option<PagingContinuation> = None;
        let mut watchdog = 0;
        loop {
            let rs_manual = session
                .execute_paged(&prepared_paged, &[], paging_continuation)
                .await
                .unwrap();
            results_from_manual_paging.append(&mut rs_manual.rows.unwrap());
            if watchdog > 30 || rs_manual.paging_state.finished() {
                break;
            }
            watchdog += 1;
            paging_continuation = rs_manual.paging_state.into_paging_continuation_opt();
        }
```

if user forgot about checking `rs_manual.paging_state.finished() {` they would go into infinite loop.

Ideas:
- Rename `PagingState` to `PagingResponse` or `PagingStateResponse` - it better conveys the semantics that this is something received from server.
- Then rename `PagingContinuation` to `PagingState` - this is the type that user will be using most (possibly storing it somewhere etc), so it's good to have a name that will be familiar from Scylla docs / other drivers.
- In next points I'll refer to those structures by the names I proposed above.
- `PagingState` could internally be an enum, or an Option, then our methods would accept `PagingState` instead of `Option<PagingState>`.
- Get rid of `into_paging_continuation_opt`
- Nit: `new_from_raw_bytes` could accept an `Arc`. If user already has an Arc then it avoids allocation. If they don't then creating it is not a big overhead.
- Rename paging_state field on QueryResult to paging_response (doesn't matter, will be removed in future commits anyway, right?)

To sum up (I didn't compile this, it's just a rough idea):
```rust
pub enum PagingStateReponse {
    NoMorePages,
    HasMorePages(PagingState)
}

pub struct PagingState {
    Start,
    Continue(Arc<[u8]>)
}

impl PagingState {
    pub fn new() -> PagingState {// or start() 
        PagingState::Start
    }

    #[inline]
    pub(crate) fn new_from_raw_bytes(raw_bytes: &[u8]) -> Self {
        if raw_bytes.len() == 0 {
             PagingState::Start
        } else {
            PagingState::Continue(Arc::from(raw_bytes))
    }

    #[inline]
    pub fn as_byte_slice(&self) -> &[u8] {
        match self {
            PagingState::Start => &[],
            PagingState::Continue(b) => b.deref()
        }
    }
}

pub async fn query_paged(
        &self,
        query: impl Into<Query>,
        values: impl SerializeRow,
        paging_continuation: PagingContinuation,
    ) -> Result<QueryResult, QueryError> {
```


Then the example before would look like this (or similar):
```rust
        let mut paging_state: PagingState = PagingState::start();
        let mut watchdog = 0;
        loop {
            let rs_manual = session
                .execute_paged(&prepared_paged, &[], paging_state)
                .await
                .unwrap();
            results_from_manual_paging.append(&mut rs_manual.rows.unwrap());
            if watchdog > 30 {
                break;
            }
            watchdog += 1;
            match  rs_manual.paging_response { // or into_paging_control_flow, not important here
                PagingResponse::NoMorePages => break,
                PagingResponse::HasMorePages(new_state) => paging_state = new_state,
            }
        }
```

While it may be a line or to longer, I think this version is cleaner and more obvious to readers.